### PR TITLE
chore(RHTAPWATCH-351): segment-bridge component deployment

### PR DESCRIPTION
**Pull request by Omeramsc**
Deploy a component for the User Journey effort as segment-bridge. This is needed in order to have a Service account with read permissions to the UserSignup resources from the host cluster. Those resources will be used in a script that'll  map between cluster usernames and SSO ID's

### CODE REVIEW

File: argo-cd-apps/overlays/development/delete-applications.yaml [hunk 1]
```diff
@@ -34,3 +34,10 @@ kind: ApplicationSet
 metadata:
   name: cluster-secret-store
 $patch: delete
+---
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: segment-bridge
+$patch: delete
```
Comment (yftacherzog) [inline argo-cd-apps/overlays/development/delete-applications.yaml:42]: Does this mean that it will not be deployed in dev? If so, do we need to maintain the dev configs under the `segment-bridge` component?

File: components/segment-bridge/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+---
+title: SEGMENT-BRIDGE
+---
```
Comment (yftacherzog) [inline components/segment-bridge/README.md:3]:
```suggestion
# SEGMENT-BRIDGE
```

File: argo-cd-apps/overlays/development/delete-applications.yaml (hunk 1 above)
Comment (Omeramsc) [reply argo-cd-apps/overlays/development/delete-applications.yaml:42]: While this is a mistake I'll fix, The current requirement is for the staging cluster, so if we'll start not including this in envs, prod should be selected as well.

Comment (yftacherzog) [reply argo-cd-apps/overlays/development/delete-applications.yaml:42]: ok. Do you want it on all stage clusters? I thought you mentioned you want it on a specific cluster.

Comment (Omeramsc) [reply argo-cd-apps/overlays/development/delete-applications.yaml]: Staging host cluster is what we need. is differentiation possible in those files?

Comment (yftacherzog) [reply argo-cd-apps/overlays/development/delete-applications.yaml]: Yes. There are ways to differentiate between the different clusters, but as far as I understand, the starting point is the collection of clusters it will be deployed on. I think currently, you're deploying only to member clusters.

Have a look here: https://github.com/redhat-appstudio/infra-deployments/tree/main/argo-cd-apps/base

Your configs are under `member`. I think it means that it will only be deployed to member clusters. You can either have it under `all-clusters` and then differentiate by name like we did for the prometheus configs, or you can have it under `host` and then I think it will only go to the host cluster.

File: argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: segment-bridge
+spec:
+  generators:
```
Comment (ifireball) [inline argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge.yaml:6]: seems like you can simplify this to just the `clusters` generator

File: components/segment-bridge/README.md [hunk 4]
```diff
@@ -0,0 +1 @@
+# SEGMENT-BRIDGE
```
Comment (ifireball) [inline components/segment-bridge/README.md:1]: At least refer to our repo so people know what his is, you can also write a few words...

File: components/segment-bridge/base/role.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
```
Comment (ifireball) [inline components/segment-bridge/base/role.yaml:2]: This gives us cluster-wide permissions, while we only need permissions for a particular namespace.

File: components/segment-bridge/base/rolebinding.yaml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: sa-read-permissions
+  namespace: segment-bridge
```
Comment (ifireball) [inline components/segment-bridge/base/rolebinding.yaml:6]: Not seeing namespace creation YAML for this namespace and we don't need our own namspace anyway, we can just put the account in the `toolchain-host-operator` namespace.

**Review by ifireball — CHANGES_REQUESTED** [review]

File: components/segment-bridge/README.md (hunk 2 above)
Comment (Omeramsc) [reply components/segment-bridge/README.md]: fixed

File: components/segment-bridge/README.md (hunk 4 above)
Comment (Omeramsc) [reply components/segment-bridge/README.md]: fixed

File: argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge.yaml (hunk 3 above)
Comment (Omeramsc) [reply argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge.yaml]: Not too sure about this actually, most of the configs looks the same. does removing the `merge` or `lists` add anything?

File: components/segment-bridge/base/role.yaml (hunk 5 above)
Comment (Omeramsc) [reply components/segment-bridge/base/role.yaml]: fixed

File: components/segment-bridge/base/rolebinding.yaml (hunk 6 above)
Comment (Omeramsc) [reply components/segment-bridge/base/rolebinding.yaml]: fixed

File: argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge.yaml (hunk 3 above)
Comment (ifireball) [reply argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge.yaml]: it would make the code simpler

File: argo-cd-apps/base/host/segment-bridge/sprayproxy.yaml [hunk 7]
```diff
@@ -0,0 +1,41 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (ifireball) [inline argo-cd-apps/base/host/segment-bridge/sprayproxy.yaml:1]: why is this file called `sprayproxy.yaml`?

File: components/segment-bridge/base/serviceaccount.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+kind: ServiceAccount
```
Comment (ifireball) [inline components/segment-bridge/base/serviceaccount.yaml:2]: we need to do something about the token for accessing this service account. AFAIK by default we would get a random token, but we need something a bit less random so we can can configure it for our client script and not have to replace it every time the cluster gets reprovisioned

**Review by ifireball — DISMISSED** [review]

File: components/segment-bridge/README.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
+
+Component used to allow reading resources from the host cluster
+
+Serve as part of the User Journey effort,
```
Comment (yftacherzog) [inline components/segment-bridge/README.md:5]:
```suggestion
Serves as part of the User Journey effort,
```

File: components/segment-bridge/base/serviceaccount.yaml (hunk 8 above)
Comment (Omeramsc) [reply components/segment-bridge/base/serviceaccount.yaml:2]: I'm looking up if we can somehow set this beforehand, maybe by setting a token secret? If not, will we need to run an oc command each time?

File: components/segment-bridge/README.md (hunk 9 above)
Comment (Omeramsc) [reply components/segment-bridge/README.md]: fixed

File: argo-cd-apps/base/host/segment-bridge/sprayproxy.yaml (hunk 7 above)
Comment (Omeramsc) [reply argo-cd-apps/base/host/segment-bridge/sprayproxy.yaml:1]: fixed.

**Comment by Omeramsc** [comment]
Added a secret to create a long-lived token. While still depends on k8s generating the token instead of pre-setting it up, its better than what we currently have

File: argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge.yaml (hunk 3 above)
Comment (Omeramsc) [reply argo-cd-apps/base/member/infra-deployments/segment-bridge/segment-bridge.yaml]: done

File: components/segment-bridge/base/serviceaccount.yaml (hunk 8 above)
Comment (Omeramsc) [reply components/segment-bridge/base/serviceaccount.yaml:2]: As mentioned in previous channels, currently I couldn't find a way to pre-set the token, and I believe it is not doable. I've set a secret-access-token so the generated token will at least be long-lived.

File: components/segment-bridge/development/kustomization.yaml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+- ../base
```
Comment (gbenhaim) [inline components/segment-bridge/development/kustomization.yaml:4]: I don't think that the sandbox operator are installed by default in development. IIRC the user should specify he/she wants to install them using a flag.

**Review by gbenhaim — CHANGES_REQUESTED** [review]

File: components/segment-bridge/base/service-account-token.yaml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: segment-bridge
+  namespace: toolchain-host-operator
```
[the remark this answers is not in the document]
Comment (ifireball) [reply components/segment-bridge/base/service-account-token.yaml:5]: yes

File: argo-cd-apps/base/host/segment-bridge/segment-bridge.yaml [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+kind: ApplicationSet
+metadata:
+  name: segment-bridge
+spec:
+  generators:
+    - clusters:
+        values:
+          sourceRoot: components/segment-bridge
+          environment: staging
+          clusterDir: ""
```
[the remark this answers is not in the document]
Comment (Omeramsc) [reply argo-cd-apps/base/host/segment-bridge/segment-bridge.yaml:11]: Is it or the `merge` generator essential? what is their use?

File: argo-cd-apps/base/member/infra-deployments/kustomization.yaml [hunk 13]
```diff
@@ -17,5 +17,6 @@ resources:
   - internal-services
   - image-controller
   - o11y
+  - segment-bridge
```
[the remark this answers is not in the document]
Comment (Omeramsc) [reply argo-cd-apps/base/member/infra-deployments/kustomization.yaml:20]: right, left it in by mistake. will remove

File: components/segment-bridge/base/role.yaml [hunk 14]
```diff
@@ -0,0 +1,8 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: segment-bridge-read-access
+rules:
+- apiGroups: [""]
```
[the remark this answers is not in the document]
Comment (Omeramsc) [reply components/segment-bridge/base/role.yaml:6]: as in `UserSignup`, right? will do.

File: components/segment-bridge/development/kustomization.yaml (hunk 10 above)
Comment (Omeramsc) [reply components/segment-bridge/development/kustomization.yaml:4]: I see there's a script to run on a dev env, but I don't see anything that needs to be configured in those files, or any other component that does that. @MatousJobanek Any idea if there's anything I need to add for it to be configured?

File: argo-cd-apps/base/host/segment-bridge/segment-bridge.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (Omeramsc) [reply argo-cd-apps/base/host/segment-bridge/segment-bridge.yaml:11]: Upon further inspection of the gitops docs [1][2], I don't see a need for the list generators as we don't generate parameters from a fixed list of cluster information,  and lacking the former - we don't need the merge generator.
[1] https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/
[2] https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators-Merge/

File: components/segment-bridge/development/kustomization.yaml (hunk 10 above)
Comment (Omeramsc) [reply components/segment-bridge/development/kustomization.yaml:4]: NVM my latest comment. 
While we configure the dev-env, we don't really need access to the sandbox operator as we install nothing. the only resource we really create there is the service account. I don't believe we need to change anything.

File: components/segment-bridge/base/serviceaccount.yaml [hunk 15]
```diff
@@ -0,0 +1,38 @@
+
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: segment-bridge-read-access
```
Comment (gbenhaim) [inline components/segment-bridge/base/serviceaccount.yaml:5]: missing namespace
you can set one for all the resources on the kustomization file.

**Review by gbenhaim — CHANGES_REQUESTED** [review]

File: components/segment-bridge/development/kustomization.yaml (hunk 10 above)
Comment (Omeramsc) [reply components/segment-bridge/development/kustomization.yaml:4]: @gbenhaim  you mean that the `toolchain-host-operator` namespace won't exist for the resources in the serviceaccount.yaml to be created in? Since we don't need the operator itself in development, and there is the issue I mentioned on my first comment, why not create a dummy namespace in development? unless I'm missing something about how to install it

File: components/segment-bridge/base/serviceaccount.yaml (hunk 15 above)
Comment (Omeramsc) [reply components/segment-bridge/base/serviceaccount.yaml:5]: fixed
will set in the kustomization file once the test will pass... for now I prefer it separated in case changes will be needed (e.g my former comment regarding creating the namespace in dev env)

**Comment by Omeramsc** [comment]
Since there's a chance the toolchain namespace won't be set in development (due to a flag configuration), leading to a sync issue, and as we don't need the development environment in the foreseeable future for this component - I've added the component to the delete overlay of development

File: components/segment-bridge/development/kustomization.yaml (hunk 10 above)
Comment (Omeramsc) [reply components/segment-bridge/development/kustomization.yaml:4]: Fixed by adding a delete overlay

**Review by gbenhaim — APPROVED** [review]

File: components/segment-bridge/base/serviceaccount.yaml [hunk 16]
```diff
@@ -29,9 +29,99 @@ roleRef:
...
+roleRef:
+  kind: Role
+  name: segment-bridge-read-member-sa-secret
+  apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: segment-bridge-read-access
+  namespace: toolchain-host-operator
```
Comment (Michkov) [inline components/segment-bridge/base/serviceaccount.yaml:68]: This role should not be here. I would put it into components/authentication. CRs should not be created in namespaces which are not managed by the component.

Comment (Omeramsc) [reply components/segment-bridge/base/serviceaccount.yaml:68]: In that case I should put all the `namespace: toolchain-host-operator` CRs there, right?

File: components/authentication/base/segment-bridge.yaml [hunk 17]
```diff
@@ -0,0 +1,64 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (gbenhaim) [inline components/authentication/base/segment-bridge.yaml:1]: Can you place it on a dedicated argocd application set? we are trying to reduce the manifests in the `authentication` component (you can do it in a different PR).

**Review by gbenhaim — APPROVED** [review]
